### PR TITLE
(PCP-297) On POSIX, create /var/run/puppetlabs if needed

### PR DESCRIPTION
**Pull request by parisiale**
Solaris cleans up /var/run at reboot. Daemonization relies on the
existence of the PID file directory. Here we create such directory when
running on Solaris, if necessary.

### CODE REVIEW

File: lib/src/configuration.cc [hunk 1]
```diff
@@ -683,6 +683,22 @@ void Configuration::validateAndNormalizeOtherSettings() {
         }
 
         auto pid_dir = fs::path(pid_file).parent_path().string();
+
+#ifdef __sun
```
Comment (kylog) [inline lib/src/configuration.cc]: Should we enable this logic everywhere? E.g. someone might configure a linux distro like this as well.

File: lib/src/configuration.cc [hunk 2]
```diff
@@ -683,6 +683,22 @@ void Configuration::validateAndNormalizeOtherSettings() {
...
+        // Solaris cleans up /var/run when rebooting; daemonization
+        // assumes that the PID file directory exists. If we're using
+        // the default directory, make sure it exists (PCP-297)
+        if (pid_dir == "/var/run/puppetlabs" && !fs::exists(pid_dir)) {
```
Comment (kylog) [inline lib/src/configuration.cc]: Should we create the directory if it doesn't exist even if it's not the default directory? The counter-argument to that idea would be that if someone has specified the pidfile manually but typo'd the path, they wouldn't want the directory created.

If we do restrict this logic to the default directory, can we DRY out occurences of `/var/run/puppetlabs` and use `DEFAULT_PID_FILE` above?

Comment (parisiale) [inline lib/src/configuration.cc]: @kylog yeah, I thought about that. But pxp-agent does not create any directories, apart from the spool ones; we followed such principle in the `Configuration` class. So, what I've tried to do here was to limit the case where we create the PID directory (Solaris platform AND default PID dir).

Also, @richardc was considering  a way to ensure the existence of the PID directory that would not involve pxp-agent directly.

Comment (kylog) [inline lib/src/configuration.cc]: Yeah, on sleeping on it, I agree with restricting this to just when `pid_dir == "/var/run/puppetlabs"`.

I still like the idea about having this fix in place for non-solaris os's or installs which mount /var/run on a temp filesystem (i.e. my comment above, not this comment :))

**Comment by mruzicka** [comment]
LGTM

**Comment by kylog** [comment]
:+1:

**Pull request merged**